src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,760 @@
+
+            return @alignCast(value_alignment, &set.values[way]);
+        }
+
+        /// Remove a key from the set associative cache if present.
+        pub fn remove(self: *Self, key: Key) void {
+            const set = self.associate(key);
+            const way = self.search(set, key) orelse return;
+
+            self.counts.set(set.offset + way, 0);
+        }
+
+        /// If the key is present in the set, returns the way. Otherwise returns null.
+        inline fn search(self: *Self, set: Set, key: Key) ?usize {
+            const ways = search_tags(set.tags, set.tag);
+
+            var it = BitIterator(Ways){ .bits = ways };
+            while (it.next()) |way| {
+                const count = self.counts.get(set.offset + way);
+                if (count > 0 and equal(key_from_value(set.values[way]), key)) {
+                    return way;
+                }
+            }
+
+            return null;
+        }
+
+        /// Where each set bit represents the index of a way that has the same tag.
+        const Ways = meta.Int(.unsigned, layout.ways);
+
+        inline fn search_tags(tags: *[layout.ways]Tag, tag: Tag) Ways {
+            const x: Vector(layout.ways, Tag) = tags.*;
+            const y: Vector(layout.ways, Tag) = @splat(layout.ways, tag);
+
+            const result: Vector(layout.ways, bool) = x == y;
+            return @ptrCast(*const Ways, &result).*;
+        }
+
+        pub fn put_no_clobber(self: *Self, key: Key) *align(value_alignment) Value {
+            return self.put_no_clobber_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Add a key, evicting an older entry if needed, and return a pointer to the value.
+        /// The key must not already be in the cache.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn put_no_clobber_preserve_locked(
+            self: *Self,
+            comptime Context: type,
+            comptime locked: fn (
+                Context,
+                *align(value_alignment) const Value,
+            ) callconv(.Inline) bool,
+            context: Context,
+            key: Key,
+        ) *align(value_alignment) Value {
+            const set = self.associate(key);
+
+            if (verify) assert(self.search(set, key) == null);
+
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            var way = self.clocks.get(clock_index);
+            comptime assert(math.maxInt(@TypeOf(way)) == layout.ways - 1);
+            comptime assert(@as(@TypeOf(way), math.maxInt(@TypeOf(way))) +% 1 == 0);
+
+            const clock_iterations_max = layout.ways * math.maxInt(Count);
```

Comment:
The tests don't cover the whole valid range of `clock_iterations_max`. For example, if we reduce `clock_iterations_max` by 1 or 2 then the tests pass.

Reply:
Turns out that the maximum number of iterations is actually `layout.ways * (math.maxInt(Count) - 1)`.

This is now covered by the tests.